core/src/main/java/org/apache/calcite/rel/RelCollationTraitDef.java

```diff
@@ -61,18 +65,21 @@ public RelCollation getDefault() {
 
   public RelNode convert(
       RelOptPlanner planner,
+      RelBuilder builder,
       RelNode rel,
       RelCollation toCollation,
+      Convention toConvention,
```

Comment:
Should the convention of `rel` be exact same with `toConvention`?

Reply:
Yes, unfortunately. We need a convention parameter because currently the RelTraitDef.convert() doesn't know which target convention it is. And Calcite, since convention is also a trait, we end up having a duplication here.

Reply:
rel has the exact convention info.

Reply:
You can use rel.getConvention() to get the convention info.

Reply:
This inherits the base class interface. For other traits, we need a convention so we can create corresponding physical rels. Unfortunately for convention trait def, this is redundant.

Reply:
We could move the convert logic to Convention class as you suggested. But I think that can be done in a separate change.

Reply:
Base class doesn't need this either. We can get convention from parameter `rel`. So the extra parameter is not needed.

Reply:
I see what you mean. fixed.